### ISSUE TITLE
Liquidity Mining Strategy Stable version

### ISSUE THREAD

**Issue by Nullably** [CoinAlpha/hummingbot#2932]
**Feature Suggestion**
This is second version of the beta Liquidity Mining strategy. Key features and fixes are as followings:
- Use hummingbot API to retrieve a list of mining campaign (markets)
- Update allocation as per this comment 
https://github.com/CoinAlpha/hummingbot/pull/2871#issuecomment-774848242

**Additional context**
- This is a follow up release on 
https://www.notion.so/hummingbot/Liquidity-Mining-strategy-0c6eb43ac5ba4c4685e3092f404e0492
- Feel free to add comments to this ticket on features you'd like to have in this new version

**Comment by Nullably** [comment]
New features/upgrades to the LM strategy I'd like to suggest are:
* Allowing unrelated markets, i.e. `token` is no longer needed, every market in `markets` will be traded.
* Allocation sized in USD amount. This creates a fixed budget (in USD term) for all markets.
* Reserved balance. This is essential when trading in BNB markets (for Binance).

I have implemented these features for my own use in `jack/liquidity_mining` branch, you can have a look see how it works.

**Comment by Nullably** [comment]
* Market make on [token]
* Order amount stays
* Performance calculation expressed in [token] amount

**Comment by rxlxrxsx** [comment]
- Add a new parameter to cancel orders when it hits a certain spread. When the market gets volatile the bot tends to adjust your spread wider, this may cause the bot to create orders not within the maximum spread allowed for a certain liquidity mining campaign therefore, not eligible for rewards. This can be optional, setting the value to -1 would disable the parameter completely.

**Comment by Nullably** [comment]
@rxlxrxsx Would it be better to have a max_spread parameter? By default the max spread is the max spread eligible for mining reward.

**Comment by rxlxrxsx** [comment]
I like this approach better. Just one question, will the orders adjusted because of volatility respect the value set on max_spread or the bot will proceed to adjust the order even if it exceeds the value set on max_spread?

Example:
max_spread set to 2
spread set to 1.5

Run the strategy, bot places orders with 1.5% spread. After some time, market becomes volatile and now adjusts your order to 2.5% spread. However because max_spread is set to 2, the bot placed the adjusted order with 2% spread instead.

**Comment by Nullably** [comment]
@rxlxrxsx 
> Run the strategy, bot places orders with 1.5% spread. After some time, market becomes volatile and now adjusts your order to 2.5% spread. However because max_spread is set to 2, the bot placed the adjusted order with 2% spread instead.

Yes, this should be the case if we chose to implement this new parameter. 
@mifeng What do you think?

**Comment by mifeng** [comment]
@Nullably Sure, I agree with this approach:
> Run the strategy, bot places orders with 1.5% spread. After some time, market becomes volatile and now adjusts your order to 2.5% spread. However because max_spread is set to 2, the bot placed the adjusted order with 2% spread instead.

Another suggestion: allow users to input either a single value or an array for the following parameters. If array, each value in the array is used for each market:
- spread
- max_spread
- order_amount
- target_base_pct

**Comment by dennisocana** [comment]
I suggest to also have `max_order_age` added. In the sample screenshot below, my order for ONE-BTC that's been active for more than 30 minutes is not earning liquidity mining rewards.

![Screen Shot 2021-02-21 at 11 26 57 AM](https://user-images.githubusercontent.com/50150287/108652706-bcec0900-74ff-11eb-959e-9be22306ae51.png)

**Comment by Nullably** [comment]
> @Nullably Sure, I agree with this approach:
> 
> > Run the strategy, bot places orders with 1.5% spread. After some time, market becomes volatile and now adjusts your order to 2.5% spread. However because max_spread is set to 2, the bot placed the adjusted order with 2% spread instead.
> 
> Another suggestion: allow users to input either a single value or an array for the following parameters. If array, each value in the array is used for each market:
> 
> * spread
> * max_spread
> * order_amount
> * target_base_pct

If the parameters accept array of values, how do we know for which market the value is? by its position on the `markets` parameter? E.g. if `markets` are `ONE-BTC,ALGO-BTC,ETH-BTC` and `spread` are `1,1.5,2`, the spreads are `ONE-BTC`: 1%, `ALGO-BTC`: 1.5% and `ETH-BTC`: 2%. If values mismatched (except 1 value to apply to all markets), this will cause validation error.

**Comment by Nullably** [comment]
> I suggest to also have `max_order_age` added. In the sample screenshot below, my order for ONE-BTC that's been active for more than 30 minutes is not earning liquidity mining rewards.
> 
> ![Screen Shot 2021-02-21 at 11 26 57 AM](https://user-images.githubusercontent.com/50150287/108652706-bcec0900-74ff-11eb-959e-9be22306ae51.png)

Ok, I'll include this in.

**Comment by mifeng** [comment]
> If the parameters accept array of values, how do we know for which market the value is? by its position on the `markets` parameter? E.g. if `markets` are `ONE-BTC,ALGO-BTC,ETH-BTC` and `spread` are `1,1.5,2`, the spreads are `ONE-BTC`: 1%, `ALGO-BTC`: 1.5% and `ETH-BTC`: 2%. If values mismatched (except 1 value to apply to all markets), this will cause validation error.

I was thinking by position, but I think you bring up a valid concern. Let's not do this yet - if users really need it, we'll get requests.

**Comment by Nullably** [comment]
@mifeng I think it's a good idea (to allow for multiple values), the validation error is so that the user can fix the setting before starting a strategy.

**Comment by Nullably** [comment]
I have updated the Liquidity Mining strategy PR to include max_spread, max_order_age and a new status format.
https://github.com/CoinAlpha/hummingbot/pull/2961
* `max_spread` is defaulted to -1 (to ignore this setting)
* `max_order_age` is defaulted to 3600 (1 hour) - current max on Kucoin, max order life on Binance is 12 hours.

**Comment by drumer979** [comment]
will in this version be included also "minimum_spread"? When the market gets volatile t hits the lower spread and kills order. This way there will be lot less filling orders.

**Comment by dennisocana** [comment]
@drumer979 thanks for your suggestion. I'll create a separate story for that so we can add to future releases. We already locked down version 0.37 development and target release is next week Wednesday.

**Comment by drumer979** [comment]
Ok. Thx
